### PR TITLE
CI: Test scipy/optimize/minpack

### PR DESCRIPTION
**Pull request by HarshitaKalani**
https://github.com/lfortran/lfortran/issues/2203

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -184,6 +184,34 @@ jobs:
...
+            git fetch ondrej
+            git checkout -t ondrej/minpack1
+            git checkout 9ee9208fc52ccc04d3128132888c0798e7ae677a
+            micromamba env create -f environment.yml
+            micromamba activate scipy-dev
+            git submodule update --init
+            all=$(find /home/harshita/scipy/scipy/optimize/minpack/ -name "*.f" | wc -l)
+            for src in $(find /home/harshita/scipy/scipy/optimize/minpack/ -name "*.f"); do
+                echo "Compiling $src"
+                lfortran -c $src -o ${src%.f}.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:202]:
```suggestion
                ../src/bin/lfortran -c $src -o ${src%.f}.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --all-mangling --magle-underscore
```

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -184,6 +184,34 @@ jobs:
...
+                cp ${src%.f}.o ./
+            done
+            ../src/bin/lfortran -c /home/harshita/scipy/scipy/optimize/minpack/chkder.f -o chkder.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --rtlib --all-mangling --mangle-underscore
+            clang -shared -fPIC -o ./scipy/optimize/minpack.so *.o -L/home/harshita/Desktop/lfortran/src/runtime -llfortran_runtime
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:206]:
```suggestion
            clang -shared -fPIC -o ./scipy/optimize/minpack.so *.o -L../src/runtime -llfortran_runtime
```

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -184,6 +184,34 @@ jobs:
...
+            mkdir -p $CONDA_PREFIX/lib/scipy/optimize/
+            cp scipy/optimize/minpack.so $CONDA_PREFIX/lib/scipy/optimize/
+            cp ../src/runtime/liblfortran_runtime.so $CONDA_PREFIX/lib/
+            python dev.py build
+            mkdir -p ./build-install/lib/python3.10/scipy/optimize/
+            cp scipy/optimize/minpack.so ./build-install/lib/python3.10/scipy/optimize/
+            python dev.py test -t scipy.optimize
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:213]:
```suggestion
            python dev.py test -t scipy.optimize -v
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
I left a few comments. @HarshitaKalani I added a new commit at `ondrej/minpack1` to use fixed versions in environment.yml file, commit hash: [b8ee2a865500263d17d366d7b286ee3a60abe6ec](https://github.com/certik/scipy/commit/b8ee2a865500263d17d366d7b286ee3a60abe6ec)

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -184,6 +184,34 @@ jobs:
...
 
+      - name: Test SciPy Minpack
+        shell: bash -e -x -l {0}
+        run: |
+            git clone https://github.com/scipy/scipy
+            cd scipy
+            git remote add ondrej https://github.com/certik/scipy
+            git fetch ondrej
+            git checkout -t ondrej/minpack1
+            git checkout 9ee9208fc52ccc04d3128132888c0798e7ae677a
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:195]:
```suggestion
            git checkout b8ee2a865500263d17d366d7b286ee3a60abe6ec
```

File: .github/workflows/CI.yml (hunk 3 above)
Comment (Pranavchiku) [reply .github/workflows/CI.yml:213]: We should use `-v` to know which test is failing.

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -184,6 +184,34 @@ jobs:
             cp scipy/special/specfun.so ./build-install/lib/python3.10/scipy/special/
             python dev.py test -t scipy.special -v
```
Comment (certik) [inline .github/workflows/CI.yml:185]: You mind need to `cd ..; rm -r scipy` here.

Comment (certik) [reply .github/workflows/CI.yml:185]: That should fix the current failure:
```
+ git clone https://github.com/scipy/scipy
fatal: destination path 'scipy' already exists and is not an empty directory.
Error: Process completed with exit code 128.
```

Comment (Pranavchiku) [reply .github/workflows/CI.yml:185]: Yes, perfect. Once we have this CI setup ready for minpack, we can combine SciPy build for Specfun and LFortran, and then just do different `python dev.py test -t scipy.special -v` followed by `python dev.py test -t scipy.optimize -v`

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -183,7 +183,39 @@ jobs:
...
+            git remote add ondrej https://github.com/certik/scipy
+            git fetch ondrej
+            git checkout -t ondrej/minpack1
+            git checkout 94fe6dd259224ae414121e62cb9cd3a2a46a1ce7
+            micromamba deactivate
+            micromamba env remove -n scipy-dev
+            micromamba env create -f environment.yml
+            micromamba activate scipy-dev
+            git submodule update --init
+            all=$(find /home/harshita/scipy/scipy/optimize/minpack/ -name "*.f" | wc -l)
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:203]:
```suggestion
            all=$(find./scipy/optimize/minpack/ -name "*.f" | wc -l)
```

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -183,7 +183,39 @@ jobs:
...
+            for src in $(find /home/harshita/scipy/scipy/optimize/minpack/ -name "*.f"); do
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:204]:
```suggestion
            for src in $(find ./scipy/optimize/minpack/ -name "*.f"); do
```

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -183,7 +183,39 @@ jobs:
...
+                echo "Compiling $src"
+                ../src/bin/lfortran -c $src -o ${src%.f}.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --all-mangling --mangle-underscore
+                cp ${src%.f}.o ./
+            done
+            ../src/bin/lfortran -c /home/harshita/scipy/scipy/optimize/minpack/chkder.f -o chkder.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --rtlib --all-mangling --mangle-underscore
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:209]:
```suggestion
            ../src/bin/lfortran -c ./scipy/optimize/minpack/chkder.f -o chkder.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --rtlib --all-mangling --mangle-underscore
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
I left some suggestions above, please address those.

**Comment by HarshitaKalani** [comment]
I believe the current failure in the checks will be resolved when https://github.com/lfortran/lfortran/pull/2263 will be merged.

File: .github/workflows/CI.yml [hunk 9]
```diff
@@ -183,7 +183,39 @@ jobs:
...
+            git checkout 94fe6dd259224ae414121e62cb9cd3a2a46a1ce7
+            micromamba deactivate
+            micromamba env remove -n scipy-dev
+            micromamba env create -f environment.yml
+            micromamba activate scipy-dev
+            git submodule update --init
+            all=$(find ./scipy/optimize/minpack/ -name "*.f" | wc -l)
+            for src in $(find ./scipy/optimize/minpack/ -name "*.f"); do
+                echo "Compiling $src"
+                ../src/bin/lfortran -c $src -o ${src%.f}.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --all-mangling --mangle-underscore
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:206]:
```suggestion
                ../src/bin/lfortran -c $src -o ${src%.f}.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --all-mangling --bindc-mangling --mangle-underscore
```

File: .github/workflows/CI.yml [hunk 10]
```diff
@@ -183,7 +183,39 @@ jobs:
...
+                cp ${src%.f}.o ./
+            done
+            ../src/bin/lfortran -c ./scipy/optimize/minpack/chkder.f -o chkder.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --rtlib --all-mangling --mangle-underscore
```
Comment (Pranavchiku) [inline .github/workflows/CI.yml:209]:
```suggestion
            ../src/bin/lfortran -c ./scipy/optimize/minpack/chkder.f -o chkder.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --rtlib --all-mangling --bindc-mangling --mangle-underscore
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
I will first have Minpack tested then specfun, else it will take a lot of time to get this PR in.

**Comment by certik** [comment]
Beautiful!!

Thanks @Pranavchiku and @HarshitaKalani.

I think all we need to do now is to "merge" the scipy changes for both specfun and minpack into just one branch, and then build minpack and specfun at once, and just build scipy once, and run tests once. So that the CI doesn't take a long time.

**Comment by Pranavchiku** [comment]
Nice, it runs! The only thing that troubles me is the time CI takes to test it in serial order. Let us have a new branch with changes of both minpack and specfun, and we can use a single build for both and test those independently. 

@HarshitaKalani I feel `ondrej/minpack1` branch works without any trouble, all tests pass, can you please verify that as you were stating about a test that fails. 

cc: @certik

**Comment by HarshitaKalani** [comment]
>I feel ondrej/minpack1 branch works without any trouble, all tests pass, can you please verify that, as you were stating about a test that fails.

Yes, everything is working perfect in my case too. All test cases are passing!

**Comment by certik** [comment]
What patches do you need to do to Minpack and/or SciPy tests to make them fully work with LFortran? If the answer is "none", then it is fully working.

As Pranav and I wrote above, let's figure out how to merge the minpack/specfun efforts and just maintain one scipy branch to test at the CI.

**Comment by HarshitaKalani** [comment]
>What patches do you need to do to Minpack and/or SciPy tests to make them fully work with LFortran? If the answer is "none", then it is fully working.

Yes, it it none.

 Also, I bumped Minpack to latest commit, everything works! I pushed it at https://github.com/certik/scipy/commits/minpack2

@Pranavchiku do let me know when can we merge both the packages.

**Comment by Pranavchiku** [comment]
Closing with reference to #2312.
